src/charm.py

```diff
@@ -33,9 +50,23 @@ def __init__(self, *args):
         self.framework.observe(
             self.on.spark_history_server_pebble_ready, self._on_spark_history_server_pebble_ready
         )
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.install, self._on_install)
-        self.spark_config = SparkHistoryServerConfig(self.model.config)
+        self.s3_creds_client = S3Requirer(self, S3_INTEGRATOR_REL)
+        self.framework.observe(
+            self.s3_creds_client.on.credentials_changed, self._on_s3_credential_changed
+        )
+        self.framework.observe(
+            self.on[S3_INTEGRATOR_REL].relation_joined, self._on_s3_credential_relation_joined
+        )
+        self.framework.observe(
+            self.on[S3_INTEGRATOR_REL].relation_broken, self._on_s3_credential_relation_broken
+        )
+        self.framework.observe(self.on[PEER].relation_changed, self._on_peer_relation_changed)
+        self.framework.observe(self.on[PEER].relation_departed, self._on_peer_relation_departed)
+        self.framework.observe(self.on[PEER].relation_broken, self._on_peer_relation_broken)
```

Comment:
**question** 

It is not clear to me whether we are using or not the PEER relation from the code? Do we store any data there?

Reply:
Had a discussion with Mykola in the morning, he suggested to not copy the S3 relation data into peer relation and use directly. So, the design change.

Reply:
Very well. As discussed yesterday, I also prefer this way to avoid storing duplicated data in the peer relation